caikit_nlp/modules/text_embedding/embedding.py

```diff
@@ -0,0 +1,384 @@
+        """Rerank the documents returning the most relevant top_n in order for this query.
+        Args:
+            query: str
+                Query is the source string to be compared to the text of the documents.
+            documents:  List[JsonDict]
+                Each document is a dict. The text value is used for comparison to the query.
+                If there is no text key, then _text is used and finally default is "".
+            top_n:  Optional[int]
+                Results for the top n most relevant documents will be returned.
+                If top_n is not provided or (not > 0), then all are returned.
+            return_documents:  bool
+                Default True
+                Setting to False will disable returning of the input document (index is returned).
+            return_query:  bool
+                Default True
+                Setting to False will disable returning of the query (results are in query order)
+            return_text:  bool
+                Default True
+                Setting to False will disable returning of document text string that was used.
+        Returns:
+            RerankQueryResult
+                Returns the (top_n) scores in relevance order (most relevant first).
+                The results always include a score and index which may be used to find the document
+                in the original documents list. Optionally, the results also contain the entire
+                document with its score (for use in chaining) and for convenience the query and
+                text used for comparison may be returned.
+
+        """
+
+        error.type_check(
+            "<NLP05323654E>",
+            str,
+            query=query,
+        )
+
+        results = self.run_rerank_queries(
+            queries=[query],
+            documents=documents,
+            top_n=top_n,
+            return_documents=return_documents,
+            return_queries=return_query,
+            return_text=return_text,
+        ).results
+
+        return (
+            results[0]
+            if len(results) > 0
+            else RerankQueryResult(scores=[], query=query if return_query else None)
+        )
+
+    @RerankTasks.taskmethod()
+    def run_rerank_queries(
+        self,
+        queries: List[str],
+        documents: List[JsonDict],
+        top_n: Optional[int] = None,
+        return_documents: bool = True,
+        return_queries: bool = True,
+        return_text: bool = True,
+    ) -> RerankPredictions:
+        """Rerank the documents returning the most relevant top_n in order for each of the queries.
+        Args:
+            queries: List[str]
+                Each of the queries will be compared to the text of each of the documents.
+            documents:  List[JsonDict]
+                Each document is a dict. The text value is used for comparison to the query.
+                If there is no text key, then _text is used and finally default is "".
```

Comment:
re: the empty fallback
That is kind of TBD. So I was concerned about getting fewer results than expected in some cases and thought "" was best for now, but perhaps forcing these to score=0 would be a future fix pending feedback TBD.

Today, I'm thinking docs w/o "text" or "_text" should be thrown out and you might get zero docs error or you might get one less result than expected.

I'd prefer that decision to go in a separate PR, but open for review discussion.

Reply:
Re: a Document type with those optional fields...
Because this rerank API is apparently used to pipe documents in and return them (e.g. for chaining I suppose) we need the Document to also have other arbitrary keys and values.  I'm not sure we can currently do your suggestion with these extra keys/values today for our data objects for gRPC and FastAPI (pydantic).  I don't think so.

The current code is able to deal with these flexible JsonDict objects (just lacks any structural hint about expected keys).

Maybe @prashantgupta24 has advice.